design-documents/asynchronous-import/base-extension.md

```diff
@@ -207,70 +160,88 @@ Will return list of Source that was uploaded before.
 }
 ```
 
-## Start File Import Endpoint
+### Base CSV Parser
```

Comment:
Usually in the streaming based file import process, we take the records from the data stream and process them as we are fetching, parsing and pushing those records. Is there a specific customer requirement, to separate the upload and import from each other. Because this design will inherently make the process slower, since you have to upload the file first, and then when it's all completed, only then you will start the actual import process. From customer perspective, they want to import the data into Magento as quickly as possible. 

On the other hand, since uploading might involve network latency and disk read (I/O latency), it is usually time consuming process, it is usually advisable to use streaming based approach end to end; meaning since lot of time spend waiting on I/O, CPU can be better off continue processing what it has already in-hand; so as you get records you can parse and push them simultaneously.

Since ultimate import will be based on Queue Processing, it should not cause dead-lock issues.

Reply:
We had a long and multiple discussion related on this topic.
Main idea of split process is to separate 1 actions that can be independent from each other.
We are talking about decomposition of processes and this approac have multiple profits.
1) We have own set of imported data, so we can always see which data were uploaded, when and by whom
2) Each file can be run independently, so you can restart import for a file as many times as you want, define different mapping for same file, fix some errors in your mapping or processing rules without requirements to re-upload whole file again and again
3) Its is possible, and there are a cases where same file can be used for processing different import types: for example - Stock and Price can be 2 different imports but they could be located in same file
4) Source - its a separate module, this mean you can just delete it. For Import its no matter where you received your file from, you can use 3rd party extension, you can create your own module that will place file in correct location and will create required entity in database. and it still will work. The same process works in another direction. You can upload file and then create any custom script that will process it. And in this case you will use all standard Magento security features for file valiation and upload.

I hope it makes sense :-)

Reply:
I believe we can still achieve all these points using a streaming based approach to this problem. Allow import again for a file that has been imported, by updating only import settings or format, there is no stopping in that. I have been involved in numerous data migration / ETL projects in the past, and have seen these issues many times.

Maybe we can improve this process later on after MVP, but I just want to make it clear, that this will not be performant for customers in many scenarios, and it will not be using the streaming based approach potential.

But I am okay, if you want to move forward initially with approach and test the waters.

Reply:
Agree!! One more point I forgot to mention is a complexity of such request. We already have an issue that for each file format we need own endpoint and own interfaces cause they have a custom format. Then also add to the body mapping information and body already become more complex and heavy. 
As we are providing API for that, I believe that if somebody will integrate it with own system, he can also build in a way of one request. 

I any case, agree with you that let’s try it how it will work and how the user experience will be. For now we had a good experience with this on our project. Let’s see

Reply:
Sure, I am okay. For large data-sets or files, this sequential implementation never performed at the desired rate for me in the past; but we can always improve on, this is not something fundamentally right or wrong. This is based on needs, streaming based approach end to end, is much more efficient in terms of memory footprint & cpu processing, that all modern systems are using it.